Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,207 @@
+require "dbm"
+require "json"
+
+#
+# `DatabaseCache` is a class acting as an interface to a persistent storage
+# mechanism residing in the `HOMEBREW_CACHE`
+#
+class DatabaseCache
+  # Name of the database cache file located at <HOMEBREW_CACHE>/<name>.db
+  #
+  # @return [String]
+  attr_accessor :name
+
+  # Instantiates new `DatabaseCache` object
+  #
+  # @param  [String] name
+  # @return [nil]
+  def initialize(name)
+    @name = name
+  end
+
+  # Memoized `DBM` database object with on-disk database located in the
+  # `HOMEBREW_CACHE`
+  #
+  # @return [DBM] db
+  def db
+    @db ||= DBM.open("#{HOMEBREW_CACHE}/#{name}", 0666, DBM::WRCREAT)
+  end
+
+  # Close the `DBM` database object after usage
+  #
+  # @return [nil]
+  def close
+    db.close
+  end
+end
+
+#
+# `CacheStore` is an abstract base class which provides methods to mutate and
```

Comment:
What about using a `Module` rather than an abstract base class?

Reply:
What advantages would extending a `Module` provide over extending an abstract base class? I'm relatively new to ruby, so I'm curious to know.

The way I approached it, I thought of what interface the `CacheStore` concretions should implement, and I thought the best way to represent that interface was through using an abstract base class.

Reply:
@AndrewMcBurney a `Module` is abstract by default effectively and you can include multiple modules into a class. In general building class hierarchies is less common in Ruby unless it really makes sense; I'm not saying it doesn't here but just worth considering I think if you're needing to raise not implemented exceptions.

Reply:
Ahh, gotcha - would you like me to change it to a `Module` instead?

Reply:
@AndrewMcBurney No, I'll leave it up to you, I just wanted to elaborate given you were saying you aren't super experience with Ruby.